crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -393,10 +549,146 @@ impl Drop for MemoryPool {
+    ///   region
+    slot_bytes: usize,
+    /// The number of stripes needed in the slab layout.
+    num_stripes: usize,
+}
+
+fn calculate(constraints: &SlabConstraints) -> Result<SlabLayout> {
+    let SlabConstraints {
+        max_memory_bytes,
+        num_memory_slots,
+        num_pkeys_available,
+        guard_bytes,
+        guard_before_slots,
+    } = *constraints;
+
+    // If the user specifies a guard region, we always need to allocate a
+    // `PROT_NONE` region for it before any memory slots. Recall that we can
+    // avoid bounds checks for loads and stores with immediates up to
+    // `guard_bytes`, but we rely on Wasmtime to emit bounds checks for any
+    // accesses greater than this.
+    let pre_slab_guard_bytes = if guard_before_slots { guard_bytes } else { 0 };
+
+    let (num_stripes, needed_guard_bytes) = if guard_bytes == 0 || max_memory_bytes == 0 {
+        // In the uncommon case where the memory or guard regions are empty, we
+        // will not need any stripes: we just lay out the slots back-to-back
+        // using a single stripe.
+        (1, guard_bytes)
+    } else if num_pkeys_available < 2 {
+        // If we do not have enough protection keys to stripe the memory, we do
+        // the same. We can't elide any of the guard bytes because we aren't
+        // overlapping guard regions with other stripes...
+        (1, guard_bytes)
+    } else if num_memory_slots == 0 {
+        (1, guard_bytes)
+    } else {
+        // ...but if we can create at least two stripes, we can use another
+        // stripe (i.e., a different pkey) as this slot's guard region--this
+        // reduces the guard bytes each slot has to allocate. We must make sure,
+        // though, that if the size of that other stripe(s) does not fully cover
+        // `guard_bytes`, we keep those around to prevent OOB access.
+        //
+        // We first calculate the number of stripes we need: we want to minimize
+        // this so that there is less chance of a single store running out of
+        // slots with its stripe--we need at least two, though. But this is not
+        // just an optimization; we need to handle the case when there are fewer
+        // slots than stripes. E.g., if our pool is configured with only three
+        // slots (`num_memory_slots = 3`), we will run into failures if we
+        // attempt to set up more than three stripes.
+        let needed_num_stripes =
+            guard_bytes / max_memory_bytes + usize::from(guard_bytes % max_memory_bytes != 0) + 1;
+        let num_stripes = num_pkeys_available
+            .min(needed_num_stripes)
+            .min(num_memory_slots);
+        let next_slots_overlapping_bytes = max_memory_bytes
+            .checked_mul(num_stripes - 1)
+            .unwrap_or(usize::MAX);
+        let needed_guard_bytes = guard_bytes
+            .checked_sub(next_slots_overlapping_bytes)
+            .unwrap_or(0);
+        (num_stripes, needed_guard_bytes)
```

Comment:
I might approach the calculation a bit differently here, along the lines of:

* There are two main invariants for each slot:
  * From the base pointer up to `limits.memory_pages` must be mappable as committed memory to be owned by the slot. AKA this is the minimum size of the slot.
  * The region from `limits.memory_pages * PAGE_SIZE .. static_memory_bound + guard` must all be guaranteed to be inaccessible by the current slot, AKA it must either be a guard region or a different stripe
* The maximum number of possible stripes is then `(static_memory_bound + guard) / (limits.memory_pages * PAGE_SIZE)`. For example with 4G bound, 2G guard, and 1G limit of growth, we can only use 6 stripes.
* If we have more stripes than the maximum needed, then everything is adjacent.
* If we have less stripes than the maximum, then each slot is inflated. For example with a 4G bound and a 2G guard and a 256MB limit, that's 24 maximum stripes but we can't ever have that many. If we got 8 stripes then I think each stripe is 6G/8 big, ~750M. This means each slot is 750M and will only ever use up to 256M of the beginning of it.

Whether or not that's simpler or not than this current calculation I'm not sure, but this is at least how I've been thinking of it and how the difference between the memory limit and memory bound I think should be factored in.

Reply:
The more I read your comment, the more I think this logic is essentially saying the same thing. Ignoring the fact that maybe `static_memory_bound` isn't handled yet, the other differences are minor:
- I'm using `guard_bytes` for `static_memory_bound + guard` and `max_memory_byes` for `limits.memory_pages * PAGE_SIZE`; I think we need to settle on a single variable (bikeshed the name) to contain these values to avoid the mental juggling of "oh, we mean how big the memory is by this calculation"
- we need at least 2 stripes here, see `needed_num_stripes`; `guard_bytes / max_memory_bytes` is essentially what you wrote up above, but we need to handle the case where this doesn't divide evenly (`usize::from(...)`)
- we need to minimize the number of stripes, see `num_stripes`: we may have fewer pkeys or memory slots than stripes needed.
- to handle the slot-inflation case, I calculate `needed_guard_bytes`--I'm considering this the entire region after a memory that needs to be inaccessible, a.k.a the MPK-reduced original guard size